### ISSUE TITLE
Incorrect unused_braces when a proc macro is applied to a GAT type conversion 

### ISSUE THREAD

**Issue by teor2345** [rust-lang/rust#109473]
I tried this code:

```rust
use serde_derive::{Deserialize};

const BAR: u32 = 1;

#[derive(Deserialize)]
struct Generic<const ASSOCIATED: u8> {}

#[derive(Deserialize)]
struct FooDeserializable {
    item: Generic<{ BAR as u8 }>,
}
```

Full example - reproduces locally with serde 1.0.158 and 1.0.156, but not on the playground with serde 1.0.156:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=98ec53a2a8de5c14af46d88f45feda1c

I expected to see this happen:

No warnings, or a warning with a working fix.

Instead, this happened:

An incorrect warning that the braces were not needed around `BAR as u8`.
An incorrect suggestion/fix to remove those brances.

### Meta

`rustc --version --verbose`:
```
rustc 1.68.0 (2c8cc3432 2023-03-06)
binary: rustc
commit-hash: 2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74
commit-date: 2023-03-06
host: x86_64-unknown-linux-gnu
release: 1.68.0
LLVM version: 15.0.6
```

The bug also exists in beta and nightly.

The warning itself might be a `serde_derive` bug, but the incorrect fix is a Rust bug:
https://github.com/rust-lang/rust/issues/73068
